dev/generate_requirements.py

```diff
@@ -0,0 +1,112 @@
+"""
+This script generates a pip-compatible requirements file (e.g. requirements/core-requirements.txt)
+based on a YAML dependencies specification (e.g. requirements/core-requirements.yaml). These
+requirements files are then read by `setup.py` to specify dependencies for MLflow wheel
+installation.
+"""
+
+import argparse
+import sys
+import yaml
+
+
+def parse_args(args):
+    parser = argparse.ArgumentParser(
+        description=(
+            "Generate a pip-compatible requirements.txt file from an MLflow"
+            " requirements.yaml specification"
+        )
+    )
+    parser.add_argument(
+        "--requirements-yaml-location",
+        required=True,
+        help="Local file path of the requirements.yaml specification.",
+    )
+    parser.add_argument(
+        "--requirements-txt-output-location",
+        required=True,
+        help="Local output path for the generated requirements.txt file.",
+    )
+
+    return parser.parse_args(args)
+
+
+def validate_requirements_yaml(requirements_yaml):
+    assert isinstance(
+        requirements_yaml, dict
+    ), "requirements.yaml contents must be a YAML dictionary"
+    for package_entry in requirements_yaml.values():
+        assert isinstance(package_entry, dict), (
+            "Entry in requirements.yaml does not have required dictionary"
+            f" structure: {package_entry}"
+        )
+        pip_release = package_entry.get("pip_release")
+        assert pip_release is not None and isinstance(pip_release, str), (
+            "Entry in requirements.yaml does not define a valid 'pip_release'"
+            f" string value: {package_entry}"
+        )
+        max_major_version = package_entry.get("max_major_version")
+        assert max_major_version is not None and isinstance(max_major_version, int), (
+            "Entry in requirements.yaml does not specify a valid 'max_major_version'"
+            f" integer value: {package_entry}"
+        )
+        if "minimum" in package_entry:
+            assert isinstance(package_entry["minimum"], str), (
+                "Entry in requirements.yaml contains an invalid 'minimum' version"
+                f" string specification: {package_entry}"
+            )
+        if "unsupported" in package_entry:
+            assert isinstance(package_entry["unsupported"], list) and all(
+                [
+                    isinstance(unsupported_entry, str)
+                    for unsupported_entry in package_entry["unsupported"]
+                ]
```

Comment:
```suggestion
            assert isinstance(package_entry["unsupported"], list) and all(
                
                    isinstance(unsupported_entry, str)
                    for unsupported_entry in package_entry["unsupported"]
                
```

nit

Reply:
Maybe we should enable this pylint rule:
https://pylint.pycqa.org/en/latest/user_guide/messages/refactor/use-a-generator.html

Reply:
Simplified! +1 to enabling this pylint rule as a follow-up PR